migrate/migrate.go

```diff
@@ -0,0 +1,109 @@
+package migrate
+
+import (
+	"strings"
+	"strconv"
+	"errors"
+	"io/ioutil"
+	"path/filepath"
+)
+
+// Migrator defines database-specific migrators ready to run migrations
+// against a single database.
+type Migrator interface {
+	Migrate() error
+}
+
+// ConnectionOptions define a collection of information that's required to
+// connect to any database in order to be able to run migrations.
+// This will be used by all `Migrator`s, either embedded in database-specific
+// options, or wholly.
+type ConnectionOptions struct {
+	Host     string
+	Port     int
+	User     string
+	Pass     string
+	Database string
+
+	protocol string
+}
+
+// ConnectionString concatenates the CockroachConnectionOptions down to a string,
+// applying defaults to options that were not set ready to be used in a
+// connection to the database
+func (co *ConnectionOptions) String() (string, error) {
+	// Start a string builder for the connection string.
+	var conn strings.Builder
+
+	// Ensure a protocol exists. If it's gibberish it is not of our concern,
+	// as an error will be returned later when trying to connect via a
+	// gibberish protocol.
+	if len(co.protocol) == 0 {
+		return "", errors.New("microservice-db: unrecognised protocol for database connection")
+	}
+
+	// Write the protocol
+	conn.WriteString(co.protocol)
```

Comment:
Still think this bit could be done more concisely with a cheeky `fmt.Sprintf("%s://...")` :P